aiida/tools/importexport/archive/writers.py

```diff
@@ -0,0 +1,301 @@
+
+__all__ = ('ArchiveData', 'ArchiveWriterAbstract', 'get_writer', 'WriterJsonFolder', 'WriterJsonTar', 'WriterJsonZip')
+
+
+def get_writer(file_format: str) -> Type['ArchiveWriterAbstract']:
+    """Return the available writer classes."""
+    writers = {
+        ExportFileFormat.ZIP: WriterJsonZip,
+        ExportFileFormat.TAR_GZIPPED: WriterJsonTar,
+        'folder': WriterJsonFolder,
+    }
+
+    if file_format not in writers:
+        raise ValueError(
+            f'Can only write in the formats: {tuple(writers.keys())}, please specify one for "file_format".'
+        )
+
+    return cast(Type[ArchiveWriterAbstract], writers[file_format])
+
+
+@dataclass
+class ArchiveData:
+    """Class for storing data, to export to an AiiDA archive."""
+    metadata: ArchiveMetadata
+    node_uuids: Set[str]
+    # UUID of the group -> UUIDs of the entities it contains
+    group_uuids: Dict[str, Set[str]]
+    # list of {'input': <UUID>, 'output': <UUID>, 'label': <LABEL>, 'type': <TYPE>}
+    link_uuids: List[dict]
+    # all entity data from the database, except Node extras and attributes
+    # {'ENTITY_NAME': {<Pk>: {'db_key': 'value', ...}, ...}, ...}
+    entity_data: Dict[str, Dict[int, dict]]
+    # Iterable of Node (pk, attributes, extras)
+    node_data: Iterable[Tuple[str, dict, dict]]
+
+    def __repr__(self) -> str:
+        """Return string representation."""
+        return f'ArchiveData(metadata={self.metadata},' + ','.join(
+            f'#{k}={len(v)}' for k, v in self.entity_data.items()
+        ) + ')'
+
+
+class ArchiveWriterAbstract(ABC):
+    """An abstract interface for AiiDA archive writers."""
+
+    def __init__(self, filename: str, **kwargs: Any):
+        """An archive writer
+
+        :param filename: the filename (possibly including the absolute path)
+            of the file on which to export.
+
+        """
+        # pylint: disable=unused-argument
+        self._filename = filename
+
+    @property
+    def filename(self) -> str:
+        """Return the filename to write to."""
+        return self._filename
+
+    @property
+    @abstractmethod
+    def file_format_verbose(self) -> str:
+        """The file format name."""
+
+    @property
+    @abstractmethod
+    def export_version(self) -> str:
+        """The export version."""
+
+    @abstractmethod
+    def write(self, export_data: ArchiveData, get_node_repo: Callable[[str], Folder]) -> dict:
+        """write the archive.
+
+        :param export_data: The data to export
+        :param get_node_repo: A callable: UUID -> to a folder of the node's repository
+        :returns: A dictionary of data about the write process
+
+        """
+
+
+def _write_to_json_archive(
+    folder: Union[Folder, ZipFolder], export_data: ArchiveData, export_version: str, get_node_repo: Callable[[str],
+                                                                                                             Folder]
```

Comment:
This is why I hate yapf 😬, it does such a poor job at handling long lines, which black is so much better at